### PR TITLE
feat(Ray): add naive sphere intersection

### PR DESCRIPTION
**Pull request by arthur-targaryen**

### CODE REVIEW

File: src/ray.rs [hunk 1]
```diff
@@ -16,15 +16,32 @@ impl Ray {
...
 
+    #[allow(dead_code)]
     pub fn at(&self, t: f64) -> Point3 {
         self.origin + t * self.direction
     }
 
+    fn hit_sphere(&self, center: &Point3, radius: f64) -> bool {
+        let oc = self.origin() - *center;
+
+        let a = self.direction.dot(&self.direction);
```
Comment (Hasenn) [inline src/ray.rs:32]: what's a, b and c ? do they have an interpretation ?
a looks likes the norm of direction squared, you could call it `length_squared` maybe

Comment (Hasenn) [reply src/ray.rs:32]: if the ray is normalized, wouldn't this be always 1 ?

File: src/ray.rs [hunk 2]
```diff
@@ -16,15 +16,32 @@ impl Ray {
...
+        let b = 2.0 * oc.dot(&self.direction);
+        let c = oc.dot(&oc) - radius * radius;
```
Comment (Hasenn) [inline src/ray.rs:34]: a comment explaining what's going on, what polynomial you are manipulating, or maybe a wikipedia link would be very welcome

File: src/ray.rs [hunk 3]
```diff
@@ -16,15 +16,32 @@ impl Ray {
...
+        let discriminant = b * b - 4.0 * a * c;
```
Comment (Hasenn) [inline src/ray.rs:35]: it looks like you used https://en.wikipedia.org/wiki/Line%E2%80%93sphere_intersection
if you did, put a link in there

**Review by Hasenn — CHANGES_REQUESTED** [review]
Please add some comments explaining your math (even if its just the wikipedia link)

File: src/ray.rs [hunk 4]
```diff
@@ -16,15 +16,32 @@ impl Ray {
...
+
+        discriminant > 0.0
+    }
+
     pub fn color(&self) -> Color {
+        if self.hit_sphere(&Point3::new(0.0, 0.0, -1.0), 0.5) {
```
Comment (Hasenn) [inline src/ray.rs:41]: If i'm raytracing i might want to know where the ray intersects the sphere, so maybe `hit (...) -> Option<HitData>` and `try_hit : (...) -> bool` would be a more sensible api ?

File: src/ray.rs [hunk 5]
```diff
@@ -16,15 +16,32 @@ impl Ray {
...
     pub fn direction(&self) -> Vec3 {
         self.direction
     }
 
+    #[allow(dead_code)]
     pub fn at(&self, t: f64) -> Point3 {
         self.origin + t * self.direction
     }
 
+    fn hit_sphere(&self, center: &Point3, radius: f64) -> bool {
```
[the remark this answers is not in the document]
Comment (arthur-targaryen) [reply src/ray.rs:29]: Done! I added a `Sphere` struct in 14219002dddc951f818f4af6d1595d790bbe346d.

File: src/ray.rs (hunk 4 above)
Comment (arthur-targaryen) [reply src/ray.rs:41]: Changed this in c18d97ec341a632a49a64f92d80868ebfba9e053 by adding `Hittable` trait, with `Hittable::try_hit()`.

File: src/ray.rs (hunk 3 above)
Comment (arthur-targaryen) [reply src/ray.rs:35]: I made it clearer by adding the link to the `Sphere::try_hit()` documentation (/a7358bd304011b42c29595614743492682e2006e) where this code has been moved. Should help understand `a`, `b` and `c` meanings.

**Pull request merged**